### PR TITLE
Fix Dart 3 dev CI builds

### PR DESCRIPTION
**Pull request by jakemac53**
Closes https://github.com/dart-lang/build/issues/3421

- Fix build_web_compilers test fixture sdk constraints to allow dart 3.x.
- Add a test utility to check if unsound null safety is supported (sdk version is 2.x).
  - Use this utility to omit tests for unsound null safety features.
- skip build_vm_compilers related tests on windows CI, there is some sort of issue related to crossing between drives

### CODE REVIEW

**Review by natebosch — APPROVED** [review]

File: build_runner/test/build_script_generate/build_script_generate_test.dart [hunk 1]
```diff
@@ -140,57 +140,69 @@ environment:
...
 
       final options = await findBuildScriptOptions(
           packageGraph: await PackageGraph.forPath('${d.sandbox}/a'));
       expect(options.canRunWithSoundNullSafety, isTrue);
     });
 
-    test('when the root package opts out', () async {
-      await d.dir('a', [
-        d.file('pubspec.yaml', '''
+    if (supportsUnsoundNullSafety) {
```
[the remark this answers is not in the document]
Comment (jakemac53) [reply build_runner/test/build_script_generate/build_script_generate_test.dart:153]: We could (I would make this a group), but I don't think it matches the semantics of what we want here as much?

Generally I see a skipped test as something that should eventually be fixed (we show them with yellow warning text). These just don't make sense to run ever.

[the remark this answers is not in the document]
Comment (jakemac53) [reply build_runner/test/build_script_generate/build_script_generate_test.dart:153]: Plus the `--run-skipped` flag would not make sense, these would fail.

[the remark this answers is not in the document]
Comment (natebosch) [reply build_runner/test/build_script_generate/build_script_generate_test.dart:153]: > Generally I see a skipped test as something that should eventually be fixed (we show them with yellow warning text).

The ones where we pass a String should eventually be fixed. I think the ones where we pass a bool it would be OK to consider as "valid" skips.

> Plus the `--run-skipped` flag would not make sense, these would fail (always).

That's a good point. I don't ever really run with that flag, but maybe I should every once in a while to catch stale skips.

Keeping as conditional blocks SGTM.
